File: server/src/controllers/patient.controller.js
// src/controllers/patient.controller.js
// Module transverse "Gestion des médecins" — fiche `patient`.
//
// Contrairement à `medecin` (annuaire public), la fiche patient est une
// donnée PRIVÉE : jamais de route publique ici. Toutes les routes
// exigent déjà "authentifier" (voir patient.routes.js) ; l'autorisation
// fine (patient concerné, médecin ayant un rendez-vous avec lui, ou
// admin/superadmin) est appliquée ICI, au cas par cas — même patron que
// estAutoriseSurRdv dans rendezVous.controller.js.
//
// Champs réels du modèle Patient (voir schema.prisma) :
//   patient { patient_id, utilisateur_id (unique), date_naissance,
//     rendez_vous[], ordonnances[] }

import prisma from "../lib/prisma.js";

const STATUTS_RDV = [
  "cree",
  "confirme",
  "en_attente_presence",
  "honore",
  "non_honore",
  "annule",
  "conteste",
];

function estAdmin(utilisateur) {
  return utilisateur?.role === "admin" || utilisateur?.role === "superadmin";
}

// Champs utilisateur exposés sur une fiche patient : réservé au
// titulaire du compte et à admin/superadmin (jamais à un médecin tiers,
// qui n'a pas besoin de l'email/téléphone du patient pour consulter son
// dossier de rendez-vous).
const SELECTION_UTILISATEUR_PATIENT = {
  select: {
    utilisateur_id: true,
    nom: true,
    prenom: true,
    email: true,
    telephone: true,
    pays_id: true,
    statut_compte: true,
  },
};

// Vue restreinte utilisée quand le demandeur est un médecin tiers
// (autorisé à voir QUE le patient existe et son nom, pas ses
// coordonnées) : même logique que SELECTION_UTILISATEUR_PUBLIC dans
// medecin.controller.js, en miroir.
const SELECTION_UTILISATEUR_RESTREINTE = {
  select: {
    utilisateur_id: true,
    nom: true,
    prenom: true,
  },
};

async function profilPatientCourant(utilisateurCourant) {
  if (!utilisateurCourant) return null;
  return prisma.patient.findUnique({
    where: { utilisateur_id: utilisateurCourant.utilisateur_id },
  });
}

async function profilMedecinCourant(utilisateurCourant) {
  if (!utilisateurCourant) return null;
  return prisma.medecin.findUnique({
    where: { utilisateur_id: utilisateurCourant.utilisateur_id },
  });
}

/**
 * Un médecin ne peut consulter la fiche d'un patient que s'il a (ou a
 * eu) au moins un rendez-vous avec lui — pas d'accès libre à tout
 * l'annuaire patients.
 */
async function medecinAUnRendezVousAvecPatient(medecinId, patientId) {
  const rdv = await prisma.rendezVous.findFirst({
    where: { medecin_id: medecinId, patient_id: patientId },
    select: { rdv_id: true },
  });
  return Boolean(rdv);
}

/**
 * Détermine si l'utilisateur courant a le droit de voir la fiche /
 * les rendez-vous du patient `patientId` : le patient lui-même,
 * admin/superadmin, ou un médecin ayant au moins un rendez-vous avec
 * ce patient (accès restreint, voir SELECTION_UTILISATEUR_RESTREINTE).
 * Retourne un objet { autorise, niveau } où niveau vaut "complet"
 * (patient lui-même ou admin) ou "restreint" (médecin tiers).
 */
async function autorisationSurPatient(patientId, utilisateurCourant) {
  if (estAdmin(utilisateurCourant)) return { autorise: true, niveau: "complet" };

  const patient = await profilPatientCourant(utilisateurCourant);
  if (patient && patient.patient_id === patientId) {
    return { autorise: true, niveau: "complet" };
  }

  const medecin = await profilMedecinCourant(utilisateurCourant);
  if (medecin && (await medecinAUnRendezVousAvecPatient(medecin.medecin_id, patientId))) {
    return { autorise: true, niveau: "restreint" };
  }

  return { autorise: false, niveau: null };
}

/* ===================================================================
 * Profil patient
 * =================================================================== */

/**
 * GET /api/patients/mon-profil
 * AUTHENTIFIÉ uniquement (voir patient.routes.js).
 * Récupère le profil complet du patient connecté (déduit du token),
 * avec ses informations utilisateur et un résumé de son activité
 * (nombre de rendez-vous, prochain rendez-vous à venir, nombre
 * d'ordonnances). Même rôle que GET /api/medecins/mon-profil côté
 * médecin : dédiée à l'écran "Mon profil" du patient.
 */
export async function obtenirMonProfil(req, res, next) {
  try {
    const utilisateurId = req.utilisateur?.utilisateur_id;
    if (!utilisateurId) {
      return res.status(401).json({ message: "Utilisateur non authentifié." });
    }

    const patient = await prisma.patient.findUnique({
      where: { utilisateur_id: utilisateurId },
      include: {
        utilisateur: SELECTION_UTILISATEUR_PATIENT,
      },
    });

    if (!patient) {
      return res.status(404).json({
        message: "Aucun profil patient associé à ce compte.",
      });
    }

    const maintenant = new Date();

    const [totalRendezVous, prochainRendezVous, totalOrdonnances] = await Promise.all([
      prisma.rendezVous.count({ where: { patient_id: patient.patient_id } }),
      prisma.rendezVous.findFirst({
        where: {
          patient_id: patient.patient_id,
          date_creneau: { gte: maintenant },
          statut: { in: ["cree", "confirme"] },
        },
        orderBy: { date_creneau: "asc" },
        include: {
          medecin: { include: { utilisateur: { select: { nom: true, prenom: true } } } },
        },
      }),
      prisma.ordonnance.count({ where: { patient_id: patient.patient_id } }),
    ]);

    return res.status(200).json({
      patient,
      statistiques: {
        total_rendez_vous: totalRendezVous,
        total_ordonnances: totalOrdonnances,
        prochain_rendez_vous: prochainRendezVous,
      },
    });
  } catch (err) {
    next(err);
  }
}

/**
 * GET /api/patients/:id
 * Ouvert au patient concerné, à admin/superadmin (vue complète,
 * coordonnées incluses), ou à un médecin ayant au moins un rendez-vous
 * avec ce patient (vue restreinte : nom/prénom uniquement, voir
 * autorisationSurPatient ci-dessus).
 */
export async function obtenirPatient(req, res, next) {
  try {
    const { autorise, niveau } = await autorisationSurPatient(req.params.id, req.utilisateur);
    if (!autorise) {
      return res.status(404).json({ message: "Patient introuvable." });
    }

    const patient = await prisma.patient.findUnique({
      where: { patient_id: req.params.id },
      include: {
        utilisateur:
          niveau === "complet" ? SELECTION_UTILISATEUR_PATIENT : SELECTION_UTILISATEUR_RESTREINTE,
      },
    });

    if (!patient) {
      return res.status(404).json({ message: "Patient introuvable." });
    }

    return res.status(200).json({ patient });
  } catch (err) {
    next(err);
  }
}

/* ===================================================================
 * Rendez-vous d'un patient
 * =================================================================== */

/**
 * GET /api/patients/:id/rendez-vous
 * Liste des rendez-vous du patient `:id`.
 * Filtre optionnel : ?statut=...
 * Accès : le patient lui-même, admin/superadmin, ou un médecin ayant
 * au moins un rendez-vous avec ce patient — dans ce dernier cas, seuls
 * SES PROPRES rendez-vous avec ce patient sont renvoyés (pas ceux pris
 * avec d'autres médecins), pour ne pas exposer le dossier complet du
 * patient à un tiers.
 */
export async function listerRendezVousPatient(req, res, next) {
  try {
    const patientId = req.params.id;
    const { statut } = req.query;

    if (statut && !STATUTS_RDV.includes(statut)) {
      return res.status(400).json({
        message: `statut invalide. Valeurs acceptées : ${STATUTS_RDV.join(", ")}.`,
      });
    }

    const { autorise, niveau } = await autorisationSurPatient(patientId, req.utilisateur);
    if (!autorise) {
      return res.status(404).json({ message: "Patient introuvable." });
    }

    const where = { patient_id: patientId };
    if (statut) where.statut = statut;

    if (niveau === "restreint") {
      const medecin = await profilMedecinCourant(req.utilisateur);
      where.medecin_id = medecin.medecin_id;
    }

    const rendezVous = await prisma.rendezVous.findMany({
      where,
      include: {
        medecin: { include: { utilisateur: { select: { nom: true, prenom: true } } } },
        structure: { select: { structure_id: true, nom: true } },
      },
      orderBy: { date_creneau: "desc" },
    });

    return res.status(200).json({ rendez_vous: rendezVous });
  } catch (err) {
    next(err);
  }
}

File: mobileapp/lib/controllers/patient_controller.dart
// lib/controllers/patient_controller.dart
//
// Gestion d'état (Riverpod) du module transverse "Gestion des
// médecins" — périmètre "fiche patient" (Patient), en miroir de
// patient_repository.dart (voir son en-tête) et de patient_models.dart,
// et dans le même esprit que medecin_controller.dart (version
// "annuaire médecin" de ce même module).
//
// Comme annoncé dans l'en-tête de [PatientRepository] : ce fichier
// porte TOUT l'état applicatif (chargement, erreurs, cache mémoire) de
// ce périmètre. Il ne parle jamais HTTP directement — il s'appuie
// uniquement sur [PatientRepository].
//
// Règle du token : identique à [PatientRepository], le token n'est
// JAMAIS conservé de façon persistante ici. Contrairement à
// [ListeMedecinsController] (medecin_controller.dart) qui met en cache
// un token optionnel pour une vue "enrichie", la fiche patient est une
// donnée PRIVÉE et TOUTES les routes l'exigent : le token est donc
// toujours passé explicitement à chaque appel de méthode ci-dessous,
// jamais mis en cache local. Idéalement, ce token provient d'un
// AuthController / authTokenProvider global déjà présent ailleurs dans
// l'app ; ce fichier ne le redéfinit pas et se contente de le recevoir
// en entrée.
//
// ⚠️ Périmètre volontairement identique à celui du repository : fiche
// "Mon profil" patient, consultation d'une fiche patient par id, et
// liste des rendez-vous d'un patient. Prise de rendez-vous, annulation,
// ordonnances sont hors périmètre — à traiter dans des controllers
// dédiés suivant le même patron.

import 'package:riverpod/legacy.dart';
import 'package:riverpod/riverpod.dart';

import '../models/patient_models.dart';
import '../repositories/patient_repository.dart';

/* =========================================================================
 * Dépendances partagées
 * ========================================================================= */

/// Repository ré-exposé ici pour que les widgets n'aient jamais besoin
/// d'importer patient_repository.dart directement.
///
/// [PatientRepository] parle HTTP directement (via le package `http`)
/// et ne prend pas de dépendance en paramètre : pas besoin d'
/// [ApiClient] ici.
final patientRepositoryProvider = Provider<PatientRepository>((ref) {
  return PatientRepository();
});

/* =========================================================================
 * Mon profil (GET /patients/mon-profil)
 * ========================================================================= */

/// Profil complet du patient connecté (fiche patient + statistiques
/// d'activité), pour l'écran "Mon profil" côté patient.
///
/// Réservé au titulaire du compte patient : le [token] doit toujours
/// être celui de la session courante, passé explicitement à [charger]
/// — voir la note sur le token en tête de fichier. `null` tant qu'aucun
/// chargement n'a été déclenché.
class MonProfilPatientController
    extends AsyncNotifier<MonProfilPatientResponse?> {
  @override
  MonProfilPatientResponse? build() => null;

  /// Charge (ou recharge) le profil courant. Conserve l'ancienne
  /// valeur affichée pendant le chargement (utile pour un pull-to-
  /// refresh sans écran de chargement plein).
  Future<void> charger({required String token}) async {
    state = const AsyncLoading<MonProfilPatientResponse?>()
        .copyWithPrevious(state);
    state = await AsyncValue.guard(
          () => ref.read(patientRepositoryProvider).obtenirMonProfil(token: token),
    );
  }

  /// Alias explicite de [charger], pour un bouton "Actualiser" /
  /// pull-to-refresh — évite à l'appelant de re-préciser son intention
  /// via le nom de la méthode.
  Future<void> rafraichir({required String token}) => charger(token: token);

  /// À appeler à la déconnexion, pour ne pas laisser la fiche d'un
  /// patient affichée après un changement de compte.
  void reinitialiser() => state = const AsyncData(null);
}

final monProfilPatientControllerProvider =
AsyncNotifierProvider<MonProfilPatientController, MonProfilPatientResponse?>(
  MonProfilPatientController.new,
);

/// Raccourci pratique pour l'écran "Mon profil" : le prochain
/// rendez-vous à venir du patient connecté (`null` si aucun rendez-
/// vous à venir, ou si le profil n'a pas encore été chargé). Dérivé de
/// [monProfilPatientControllerProvider] plutôt que recalculé côté UI.
final prochainRendezVousProvider = Provider.autoDispose<RendezVousPatient?>((ref) {
  final profil = ref.watch(monProfilPatientControllerProvider).asData?.value;
  return profil?.statistiques.prochainRendezVous;
});

/* =========================================================================
 * Fiche patient par id (GET /patients/:id)
 * ========================================================================= */

/// Fiche d'un patient par son id — vue complète (patient lui-même,
/// admin/superadmin) ou restreinte (médecin tiers ayant un rendez-vous
/// avec ce patient), voir [UtilisateurPatient.estVueComplete].
///
/// `autoDispose` : pas de raison de garder une fiche consultée en
/// mémoire une fois l'écran de détail quitté (donnée privée).
/// `family` : une fiche par couple (id, token) — deux appelants avec
/// des droits différents sur le même patient (ex. le patient lui-même
/// puis un médecin) obtiennent chacun leur propre requête/cache, sans
/// jamais mélanger vue complète et vue restreinte.
final patientParIdProvider = FutureProvider.autoDispose
    .family<Patient, ({String id, String token})>((ref, params) {
  return ref
      .read(patientRepositoryProvider)
      .obtenirPatient(params.id, token: params.token);
});

/* =========================================================================
 * Rendez-vous d'un patient (GET /patients/:id/rendez-vous)
 * ========================================================================= */

/// Filtre de statut courant pour l'écran "rendez-vous d'un patient" —
/// à faire varier (ex. depuis des chips "à venir / honorés / annulés")
/// puis relire via [rendezVousPatientProvider] avec la même valeur.
/// `null` = aucun filtre, tous statuts confondus.
final filtreStatutRendezVousProvider =
StateProvider.autoDispose<StatutRendezVous?>((ref) => null);

/// Liste des rendez-vous du patient [id], du plus récent au plus
/// ancien, filtrée par [statut] si fourni.
///
/// Mêmes règles d'accès que [patientParIdProvider] ; pour un médecin
/// tiers autorisé, seuls SES PROPRES rendez-vous avec ce patient sont
/// renvoyés (jamais le dossier complet) — restriction appliquée côté
/// backend, ce controller ne fait que relayer la réponse.
///
/// `family` : une liste par combinaison (patientId, token, statut).
/// Pour recharger après un changement de [filtreStatutRendezVousProvider],
/// il suffit que le widget appelant relise ce provider avec la nouvelle
/// valeur de statut — Riverpod se charge de refaire la requête et de
/// garder les anciennes combinaisons en cache le temps de l'autoDispose.
final rendezVousPatientProvider = FutureProvider.autoDispose.family<
    List<RendezVousPatient>,
    ({String patientId, String token, StatutRendezVous? statut})>((ref, params) {
  return ref.read(patientRepositoryProvider).listerRendezVousPatient(
    params.patientId,
    token: params.token,
    statut: params.statut,
  );
});

/// Variante pratique de [rendezVousPatientProvider] ne gardant que les
/// rendez-vous à venir (voir [RendezVousPatient.estAVenir]), pour un
/// onglet "à venir" sans dupliquer la logique de filtrage côté widget.
/// Ne fait pas d'appel réseau supplémentaire : dérive simplement le
/// résultat déjà chargé sans filtre de statut.
final rendezVousAVenirPatientProvider = FutureProvider.autoDispose
    .family<List<RendezVousPatient>, ({String patientId, String token})>(
        (ref, params) async {
      final tous = await ref.watch(rendezVousPatientProvider(
        (patientId: params.patientId, token: params.token, statut: null),
      ).future);
      return tous.where((rdv) => rdv.estAVenir).toList();
    });